iroha/src/p2p.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    pub async fn start_and_connect(&self, peer_address: &str) -> Result<(), String> {
+        let peer = Peer {
+            listen_address: peer_address.to_string(),
+        };
+        self.state.lock().await.peers.insert(peer.clone());
+        let message = Message::NewPeer(Peer {
+            listen_address: self.state.lock().await.listen_address.clone(),
+        });
+        Self::send(&message, peer).await?;
+        self.start().await?;
+        Ok(())
+    }
+
+    async fn send(message: &Message, peer: Peer) -> Result<(), String> {
+        let _response = Network::send_request_to(
+            peer.listen_address.as_ref(),
+            Request::new("/".to_string(), message.into()),
+        )?;
+        Ok(())
+    }
+
+    async fn broadcast(state: State<NodeState>, message: &Message) -> Result<(), String> {
+        for peer in state.lock().await.peers.clone() {
+            Self::send(message, peer).await?;
+        }
+        Ok(())
+    }
+
+    async fn ping(state: State<NodeState>, peer: Peer) -> Result<(), String> {
+        let ping = Ping {
+            payload: gen_random_bytes(PING_SIZE),
+            to_peer: peer.clone(),
+            from_peer: Peer {
+                listen_address: state.lock().await.listen_address.clone(),
+            },
+        };
+        state
+            .lock()
+            .await
+            .sent_pings
+            .insert(ping.clone(), current_time());
+        Self::send(&Message::Ping(ping), peer.clone()).await
+    }
+
+    async fn ping_all(state: State<NodeState>) -> Result<(), String> {
+        for peer in state.lock().await.peers.clone() {
+            Self::ping(state.clone(), peer).await?;
+        }
+        Ok(())
+    }
+
+    async fn handle_message(state: State<NodeState>, bytes: &[u8]) -> Result<(), String> {
+        let message: Message = bytes.to_vec().into();
+        match message {
+            Message::Ping(ping) => {
+                Self::send(&Message::Pong(ping.clone()), ping.from_peer).await?;
+            }
+            Message::Pong(ping) => {
+                let sent_pings = &mut state.lock().await.sent_pings;
+                if sent_pings.contains_key(&ping) {
+                    let sent_time = sent_pings
+                        .get(&ping)
+                        .expect("Failed to get sent ping entry.");
+                    let _rtt = current_time() - sent_time.to_owned();
+                    sent_pings.remove(&ping);
+                }
+            }
+            Message::PendingTx(_tx) => {
+                //TODO: handle incoming pending tx
+            }
+            Message::NewPeer(peer) => {
+                //TODO: use transactions to add a new peer and verify on connection in swarm
+                //tell node about other peers
+                for peer in state.lock().await.peers.clone() {
+                    let message = Message::AddPeer(peer.clone());
+                    Self::send(&message, peer.clone()).await?
```

Comment:
the same here, it's better because it's inlined, but let's spawn features in parallel